freeglut/freeglut/src/fg_internal.h

```diff
@@ -526,11 +579,14 @@ typedef void (*SFG_Proc)();
  * and for no other reason.  Since it's hidden in the macro, the
  * ugliness is felt to be rather benign.
  */
-#define SET_WCB(window,cbname,func)                            \
+#define SET_WCB(window,cbname,func,udata)                      \
 do                                                             \
 {                                                              \
     if( FETCH_WCB( window, cbname ) != (SFG_Proc)(func) )      \
```

Comment:
I didn't see a reason why the developer couldn't re-provide the function (nearly all window callbacks are just set functions anyway), nor a reason why the developer couldn't provide a struct pointer of some sort that could have it's contents modified.

I tried to keep all the functions consistent, and had written the window callbacks last. I didn't see a good place for many other callback functions to have an additional set operation, and since window callback data is per-callback, it could result in a bunch of `glut` functions who's implementation would all be:
```
void glutSetXXXUserData(void* user_data)
{
    SET_WCB_USER_DATA(XXX, user_data);
}
```

Not to say it can't be done, but I'm not sure why I should pollute the `glut` namespace more when the callbacks can be re-set to the window with new data.

I'll take the opportunity to note that near the end of my initial implementation of the window callbacks, I noticed an already implemented user data parameter for windows and menus (`glutSetWindowData`/`glutGetWindowData`, `glutSetMenuData`/`glutGetMenuData`). It was per-window/per-menu, as opposed to this per-callback version, and seems to be a freeglut implementation as opposed to Glut. At the time I had a "oh come on, someone already did this?!" but given they were broad and weren't in "the unwritten standard for callback user data" format, I questioned how many people knew about it and used it. While I probably can't use it within the implementation I provided, I'm not sure if anything should be done with it or if it should be left as is. For all I know, someone will say "skip your implementation and just do that". Understandable, but devs (myself included) tend to look straight ahead and not look what other functions exist unless it's a new library. Suddenly adding callbacks that take user data will be noticed, even when something similar may have existed for a decade or more. So will more of the same get people to use it?